### PR TITLE
Fix clippy warnings

### LINKED ISSUE #10: Fixing clippy warnings
- [ ] Fix warnings
- [ ] Gate clippy on CI

### PR DESCRIPTION
**Pull request by ArifRoktim**
Should close #10

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -45,3 +45,11 @@ jobs:
           rustup toolchain install 1.36.0
           rustup default 1.36.0
       - run: cargo build --all-targets --all
+
+  clippy:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - run: rustup component add clippy
+      - run: cargo clippy
```
Comment (lzutao) [inline .github/workflows/main.yml:54]:
```suggestion
      - run: cargo clippy --all --all-targets
```

**Comment by ArifRoktim** [comment]
Clippy on latest stable suggests replacing lines like 
```rust
self.executable_dir.as_ref().map(|p| p.as_path())
```
with 
```rust
self.executable_dir.as_deref()
```
`Option::as_deref` was introduced in 1.40.0. I'm thinking about adding 
`#![allow(clippy::option_as_ref_deref)]` with a comment to remove it when msrv is bumped past 1.40.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -45,3 +45,10 @@ jobs:
           rustup toolchain install 1.36.0
           rustup default 1.36.0
       - run: cargo build --all-targets --all
+
+  clippy:
+    runs-on: ubuntu-latest
```
Comment (lzutao) [inline .github/workflows/main.yml:50]: Could you use Actions matrix to run clippy on windows and macOS too?

**Comment by ArifRoktim** [comment]
Different clippy versions were tested by running:
```sh
mkdir clippy_versions
for i in 1.{34..42}.0 stable; do
    find -name '*.rs' -execdir touch {} \;
    cargo +${i} clippy --all --all-targets &> clippy_versions/clippy-${i}.txt
done
```

---
`clippy::option_as_ref_deref`, introduced with the clippy that comes with rust 1.42.0, is not compatible with the MSRV until the MSRV is bumped past 1.40.0, so I ignore the lint.
But clippy from rust MSRV until rust 1.41.0 will complain about the lint being unknown, so I also ignore `clippy::unknown_clippy_lints`.

---
`clippy::redundant_closure` in rust 1.34.0 was overzealous and included warnings for closures containing method calls. Since 1.35.0, this lint has been split into two by rust-lang/rust-clippy/pull/4101 and there is now a separate `clippy::redundant_closure_for_method_calls` that is ignored by default.

File: directories/src/lib.rs [hunk 3]
```diff
@@ -15,6 +15,17 @@
 #![deny(missing_docs)]
 #![warn(rust_2018_idioms)]
 
+// FIXME(MSRV >= 1.40.0): rust 1.40.0 stabilized `Option::as_deref`. Starting with rust 1.42.0,
```
Comment (ArifRoktim) [inline directories/src/lib.rs:18]: I wasn't sure how to format the comment to indicate that this comment and the associated `#![allow()]` should be removed when MSRV is greater than or equal to 1.40.0.
I think this way makes sense and is easy enough to understand.

**Comment by lzutao** [comment]
I suggest using commandl-ine flag to allow clippy lint.

**Comment by ArifRoktim** [comment]
Clippy on the command line doesn't seem to heed `clippy::unknown_clippy_lints`:

```sh
$ cargo +1.42.0 clippy --all --all-targets -- -A clippy::option_as_ref_deref
    Checking dirs-sys-next v0.1.0 (/home/arif/git/dirs/dirs-sys)
    Checking directories-next v1.0.0 (/home/arif/git/dirs/directories)
    Checking dirs-next v1.0.0 (/home/arif/git/dirs)
    Finished dev [unoptimized + debuginfo] target(s) in 0.45s

$ cargo +1.41.0 clippy --all --all-targets -- -A clippy::option_as_ref_deref
    Checking dirs-sys-next v0.1.0 (/home/arif/git/dirs/dirs-sys)
error[E0602]: unknown lint: `clippy::option_as_ref_deref`
  |
  = help: did you mean: `clippy::option_unwrap_used`
  = note: requested on the command line with `-A clippy::option_as_ref_deref`

error[E0602]: unknown lint: `clippy::option_as_ref_deref`
  |
  = help: did you mean: `clippy::option_unwrap_used`
  = note: requested on the command line with `-A clippy::option_as_ref_deref`

error: aborting due to previous error
<snip>

$ cargo +1.41.0 clippy --all --all-targets -- -A clippy::unknown_clippy_lints -A clippy::option_as_ref_deref
    Checking dirs-sys-next v0.1.0 (/home/arif/git/dirs/dirs-sys)
error[E0602]: unknown lint: `clippy::option_as_ref_deref`
  |
  = help: did you mean: `clippy::option_unwrap_used`
  = note: requested on the command line with `-A clippy::option_as_ref_deref`

error[E0602]: unknown lint: `clippy::option_as_ref_deref`
  |
  = help: did you mean: `clippy::option_unwrap_used`
  = note: requested on the command line with `-A clippy::option_as_ref_deref`

error: aborting due to previous error
<snip>
```
Plus if in the code itself, the output when users run clippy in the repo will be the same as that of the CI.

**Comment by lzutao** [comment]
How about using clippy of nightly toolchain?

**Comment by ArifRoktim** [comment]
Whoops, I think I misunderstood what you wanted. I was trying to make it so that users who ran clippy on the shell using any version from the MSRV to the current stable didn't get extraneous warnings, which is why I added the `#![allow(clippy::<lint>)]` in the source files.

I think you're only concerned about the CI. If that's the case, then the latest stable is all that's needed. All of `os: [ubuntu-latest, macos-latest, windows-latest]` use 1.43.0+, so it should be fine to just do:
`cargo clippy --all --all-targets -- -A clippy::option_as_ref_deref`.

**Comment by lzutao** [comment]
Yes, I don't think it is useful to use old clippy for linting.

**Comment by lzutao** [comment]
bors has often crashed recently. Merge manually I think.

**Pull request merged**